SCUMM: Some graphics rendering cleanup

beeanyew opened the pull request:
While working on the AmigaOS 3 port of ScummVM, I ran into issues
with a lot of spread out code and looparounds, making it difficult to
add any optimizations that would allow the m68k processors in these
machines to run things better, so I wanted to clean up at least a bit
of the main graphics rendering loop.

There are still quite a few different memcpy/rendering loops in the
other files (charset, actor, etc.), but I will continue to look at
these as I make progress with getting the AmigaOS 3 port back
in better shape.

This lowers CPU load in games that do a lot of blast object/text
rendering, such as Curse of Monkey Island (naturally) and the
V6/V7 games by avoiding data copying in favor of just erasing dirty
rects from layers directly. The simple two layer rendering for
games like INDY4 may be a tiny bit slower than before, but not to
any noticeable extent even on an m68k processor.

NOTE:
I have not tested all SCUMM/HE games with these changes, but I did
test most of the ones I own to a reasonable extent to make sure that
everything seemed to be working correctly compared to before these
changes were made.)

Mataniko commented:
Very cool, do you have some metrics around impact to CPU usage for these changes?

beeanyew commented:
I don't have any precise metrics, unfortunately. The V6/V7 games all hovered around 0% CPU usage on my main PC, any difference was too small for even `clock()` (microseconds) to measure it. CoMI dropped in CPU usage from spikes to ~3% total CPU usage to around 1%. There may or may not be a very minor negative performance impact on pre-V6 games, but if there is, it's not enough to be noticeable even on a 50MHz CPU. The V6+ games all ran better when I tested them on my Amiga 4000 compared to before, CoMI showing (in a relative sense, of course) amazing improvements in frame times overall.

sev- commented on engines/scumm/actor.cpp:
```diff
@@ -2209,18 +2209,35 @@ void ScummEngine_v90he::processActors() {
+void ScummEngine::removeUpperActors() {
+	int i;
+
+	for (i = 0; i < _upperActorQueuePos; i++) {
```
Please write `for (int i = 0;...` for reducing scope of the variable.

beeanyew replied:
Got it, I merely went with this approach because `processUpperActors` used it, and I was trying to stay consistent with what was already in there. :)

sev- commented on engines/scumm/gfx.cpp:
```diff
@@ -343,31 +338,53 @@ void ScummEngine::initScreens(int b, int h) {
+			_virtscr[kVerbVirtScreen].fillRect(clear_rect, CHARSET_MASK_TRANSPARENCY);
 		}
 	}
+	else {
```
This code formatting is incorrect. Should be `} else {`. Same thing above.

sev- commented on engines/scumm/gfx.cpp:
```diff
@@ -513,57 +565,50 @@ void ScummEngine::drawDirtyScreenParts() {
+		_layers[0] = (!_blastObjectQueuePos) ? NULL : &_virtscr[kUnkVirtScreen];
+		_layers[1] = NULL;
+		_layers[3] = (!_blastTextQueuePos) ? NULL : &_virtscr[kTextVirtScreen];
+	}
```
Code formatting. `} else {`

sev- commented on engines/scumm/gfx.cpp:
```diff
@@ -1109,30 +1173,44 @@ void ScummEngine::restoreCharsetBg() {
-				// Restore from back buffer
-				const byte *backBuf = vs->getBackPixels(0, 0);
-				blit(screenBuf, vs->pitch, backBuf, vs->pitch, vs->w, vs->h, vs->format.bytesPerPixel);
+		for (i = 0; i < _gdi->_numStrips; i++) {
```
Please reduce scope of `i` variable.

sev- commented on engines/scumm/gfx.cpp:
```diff
@@ -1109,30 +1173,44 @@ void ScummEngine::restoreCharsetBg() {
-		// restoreBackground(), but was changed to only restore those parts which are
-		// currently covered by the charset mask.
+		int i;
+		int w = 8;
```
Seems that `w` is used only within `for (i = 0; i < _gdi->_numStrips; i++) {` loop. Please move it there for reducing its scope.

sev- commented on engines/scumm/gfx.h:
```diff
@@ -69,10 +69,12 @@ struct CameraData {
-	kTextVirtScreen = 1,	// In V0-V3 games: the area where text is printed
-	kVerbVirtScreen = 2,	// The verb area
-	kUnkVirtScreen = 3		// ?? Not sure what this one is good for...
+	kMainVirtScreen,	// The 'stage'
```
Please return the enum values back, because they're used by the scripts.

beeanyew replied:
Ah, these (integer) values are actually the same as before, they're just not explicitly stated since it seemed redundant to do so, but I can restore the explicit value assignments no problem.

sev- commented on engines/scumm/gfx.h:
```diff
@@ -156,6 +158,12 @@ struct VirtScreen : Graphics::Surface {
+	 * The size of the allocated memory area in bytes.
+	 * Mostly useful to avoid unnecessary re-allocations of the buffer.
+	 */
+	int byte_size;
```
Maybe it is better to name it `bufferSize` then?

sev- reviewed (changes requested):
I browsed through your changes for the obvious things and code formatting issues.

Will perform a deeper analysis later.

sev- commented:
Any reason why did you close this?

beeanyew commented:
Uhh, I was cleaning up some of my old github stuff earlier, and figured that since there had been no activity here for months that it wasn't interesting for the main ScummVM branch. The changes themselves are still available in my scummvm-amigaos3 branch.

samo79 commented:
@beeanyew

I believe it will worth merging as your modifications could also be useful for those who use ScummVM on slow computers running OS4 and MOS

beeanyew commented:
Hm. It does sound rather unlikely, considering that even the slowest PPC running MorphOS or AmigaOS 4 should be multiples of tens times the speed of say a 50MHz 68030, or even a 50MHz 68060 with a massively higher memory bandwidth. If they're having issues updating a 320x200 screen, the problems are probably somewhere else in their Platform implementation.

samo79 commented:
Yep maybe, however there are multiple reasons to have that changes included also in main repository .. also for others who eventually in the future would want to continue your port on AmigaOS 3.x

beeanyew commented:
I haven't dropped the AmigaOS 3.0 port at all, which is why I'm saying that the changes are still in that repo on my github. All I've said is that the AmigaOS 3.0 port can't be made up to date with the main branch anymore due to the use of features such as UTF-32 strings and engine dependencies on libraries that are either nearly impossible to build or do not perform to any reasonable degree on this operating system from 1992. Just grab my code and make it your PR, I don't want credit for it or anything.

samo79 commented:
Well, if from the analysis of the code no problems will emerged (or any other changes that need to be made) I think we could already do the merge ..
I can test these later on AmigaOS4 for eventual regressions/issues

@sev

What do you think, can you merge the current code ?

sev- commented:
Yes, I think, the code is worth merging.

samo79 commented:
@sev- 

Any news ? :-)

sev- commented:
@samo79 what news? @beeanyew needs to restore his branch...

beeanyew commented:
Just to make this clear, there's no branch to restore. I removed the scummvm repo fork from my github account because I have no interest in the main ScummVM branch. All the code I changed is still available in the commits displayed on the page for this pull request, and the changes are also in my amigaos3 repo over at https://github.com/beeanyew/scummvm-amigaos3/tree/zz9k-stuff

Like I said, I want no credit for this, or anything to do with it. Take my code and make it your own PR. It's a bit of work, sure, but not anything impossible.

samo79 commented:
@sev- 

I meant if you had the time to merge these changes into the main repo, of course once you have verified the correctness of the changes.. eventually once merged i'm availible to test these under AmigaOS4

@beeanyew

Among classic Amiga of today there is also the Vampire with cpu 68080 ..

beeanyew commented:
Please stop tagging me. I know of the Vampire accelerator boards, and they don't need any of these changes either, because they're essentially 4x the speed of a 50MHz 68060 with an infinitely faster memory interface for their RTG.

samo79 commented:
@sev- 

Ik the topic was closed but the code was not merged yet in main repo, can you do that when you have the time?
Thanks

Mataniko commented:
@samo79 There's nothing to merge as the changes require additional work, someone will have to pick it up and submit a new PR before this can be considered.

mikrosk commented:
@beeanyew As I'm working on something similar (Atari Falcon 060@66 MHz), @samo79 was kind enough to point me to your PR.

> All I've said is that the AmigaOS 3.0 port can't be made up to date with the main branch anymore due to the use of features such as UTF-32 strings and engine dependencies on libraries that are either nearly impossible to build or do not perform to any reasonable degree on this operating system from 1992.

Maybe this was true when you were working on this PR but from my experience, basically everything is possible to switch off. I'm linking my scummvm only with zlib, nothing else, out of the box.

> the changes are also in my amigaos3 repo over at https://github.com/beeanyew/scummvm-amigaos3/tree/zz9k-stuff

Unfortunately, it seems you've removed the whole repository at some point. Any chance of bringing it back? I don't see your scummvm flavour even on aminet. :-(

beeanyew commented:
Hello, unfortunately I cannot assist you with this, if you had caught me last year, it would have been possible.

I have since that time permanently blocked and signed out of GitHub, left all development and Amiga communities and deleted all of my code locally as I no longer wish to participate in anything. (I am replying to you using email, I cannot access the web site.)

Keeping GPL licensed source code in a fashion where it is unavailable to everyone else is not something the license allows.

 

The only thing I can think of is checking the “PiStorm” and “Amiga dev” (the ScummVM channel specifically) Discord servers, see if anyone forked or cloned one of my repos, I physically no longer have any of my old files that were available exclusively on GitHub.

 

From: Miro Kropáček ***@***.***> 
Sent: Tuesday, February 7, 2023 4:15 PM
To: scummvm/scummvm ***@***.***>
Cc: beeanyew ***@***.***>; Mention ***@***.***>
Subject: Re: [scummvm/scummvm] SCUMM: Some graphics rendering cleanup (#2382)

 

@beeanyew <https://github.com/beeanyew>  As I'm working on something similar (Atari Falcon ***@***.*** MHz), @samo79 <https://github.com/samo79>  was kind enough to point me to your PR.

All I've said is that the AmigaOS 3.0 port can't be made up to date with the main branch anymore due to the use of features such as UTF-32 strings and engine dependencies on libraries that are either nearly impossible to build or do not perform to any reasonable degree on this operating system from 1992.

Maybe this was true when you were working on this PR but from my experience, basically everything is possible to switch off. I'm linking my scummvm only with zlib, nothing else, out of the box.

the changes are also in my amigaos3 repo over at https://github.com/beeanyew/scummvm-amigaos3/tree/zz9k-stuff

Unfortunately, it seems you've removed the whole repository at some point. Any chance of bringing it back? I don't see your scummvm flavour even on aminet. :-(

—
Reply to this email directly, view it on GitHub <https://github.com/scummvm/scummvm/pull/2382#issuecomment-1420942700> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AFAHSG6Q34AP7RBT2L5YPFDWWJRHZANCNFSM4PIPU6WQ> .
You are receiving this because you were mentioned.Message ID: ***@***.***>

digitall commented:
@mikrosk : There are several clone / trees available on Github for scummvm-amigaos3.

The first one is based on v1.8.1 release and was last updated in 2016:
https://github.com/mntmn/scummvm-amigaos3

The second one is more up to date, based on v2.0.0 release:
https://github.com/AmigaPorts/scummvm-amigaos3

The most up to date one I could find is based on v2.2.0 release:
https://github.com/betajaen/scummvm-amigaos3

Hope that helps.

samo790 commented:
Versions 1.8.1 and 2.0.0 for AmigaOS 3 68k were lastly updated in 2016 and 2020, so for sure they cannot include the changes of this PR that beeanyew is looking for...

The most updated fork for AmigaOS 3 68k are the one of betajean based on the 2.2.0 release.
However even this version doesn't include any of the optimization changes done by beeanyew in this PR

Aniway even if the repo was shutten down, i suggest to pick the code from here:
https://github.com/scummvm/scummvm/pull/2382/commits

samo790 commented:
@mikrosk 
That's a very good news, keep it up! :-)

mikrosk commented:
Actually, in the end I took the opposite route - to be sure that the fix/patch is really working, I backported my Atari changes to `branch-2-2`, with and without this PR.

And I have bad news -- I simply wasn't able to reproduce any significant speedup on Atari Falcon 060@66 MHz. I even tried to disable the whole rendering part (using the null renderer), no difference either.

I've been in touch with beeanyew, he could remember some explicit cases of huge speedup he'd seen (e.g. the intro credits in Full Throttle) but to no avail. There can be various reasons for this:

- beeanyew forgot to publish some crucial part of his code to enable this behaviour
- somewhere between his code and 2.2 release someone else added some code which either disabled this behaviour and/or optimised rendering to the point the fixes were not really needed
- his hardware (68030@50 MHz + some RTG card) was way more sensible to even slightest optimisations
- Atari Falcon's hardware is so much faster that even if the optimisations are there and working, it's not noticeable
- I did something stupid ;)

So I'm leaving this for some fellow Amiga coder. Fetching & creating this PR's branch is easy: `git fetch origin pull/2382/head:pr-2382` and as I said earlier, it is quite easy to rebase it on `branch-2-2` up to `branch-2-5` (after that some functions from this PR started to be deprecated and/or removed).

samo790 commented:
68060 are maybe powerfull enough to spot any differences, eventually you may try asking someone in the Atari community to try this executable with a slower 68020/68030/68040 processors

mikrosk commented:
@samo79 let's discuss the question of merging here. As mentioned, it's not so easy. For 2.2 - 2.5, sure, it is but when it comes to master, I couldn't even vouch for this code to be doing something.

And as you have seen in the Atari PR, guys here can be pretty thorough when it comes to reviews, so I'd feel pretty uncomfortable defending the patch, especially if parts of it will have to be removed due to a different way how master solves the problem in the present. Not to mention my inability to reproduce the issue in the first place.

So the best I can do for you is to offer a rebased `branch-2-5` with this patch and someone with the Amiga toolchain can compile and run it to see whether is some difference.
